// @rork-ai/toolkit-sdk/src/integrations 2.ts
import { useMutation, useQuery } from "@tanstack/react-query";
import * as WebBrowser from "expo-web-browser";

import { withBaseUrl } from "./api";

type Connection = {
  id: string;
  status:
    | "INITIALIZING"
    | "INITIATED"
    | "ACTIVE"
    | "FAILED"
    | "EXPIRED"
    | "INACTIVE";
  toolkit: string;
};

const EMPTY_ARRAY = [] as Connection[];

async function waitForConnection(id: string) {
  const response = await fetch(withBaseUrl(`/connection/wait/${id}`));
  return response.json() as Promise<{ status: string }>;
}

export function useConnections() {
  const list = useQuery({
    queryKey: ["rork__connections"],
    queryFn: async () => {
      const response = await fetch(withBaseUrl("/connections/list"));
      return response.json() as Promise<{ connections: Connection[] }>;
    },
  });

  const initiate = useMutation({
    mutationFn: async (toolkit: string) => {
      const authUrl = withBaseUrl(`/connection/init/${toolkit}`);
      const response = await fetch(authUrl).then((res) => res.json());
      const data = response as {
        id: string;
        status: string;
        redirectUrl: string;
      };

      const resultPromise = WebBrowser.openAuthSessionAsync(
        data.redirectUrl,
        "demotoolkit://connection/callback",
      );

      const result = await resultPromise;

      if (result.type === "success") {
        list.refetch();
      }

      return result;
    },
  });

  const disconnect = useMutation({
    mutationFn: async (toolkit: string) => {
      const response = await fetch(
        withBaseUrl(`/connection/disconnect/${toolkit}`),
      );
      return response.json() as Promise<{ status: string }>;
    },
  });

  return {
    currentConnections: list.data?.connections ?? EMPTY_ARRAY,
    refetchConnections: list.refetch,
    initiate,
    disconnect,
  };
}
